// @govtechsg/oa-verify/dist/cjs/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var verificationBuilder_1 = require("./verifiers/verificationBuilder");
exports.verificationBuilder = verificationBuilder_1.verificationBuilder;
var openAttestationHash_1 = require("./verifiers/hash/openAttestationHash");
exports.openAttestationHash = openAttestationHash_1.openAttestationHash;
var openAttestationDnsTxt_1 = require("./verifiers/dnsText/openAttestationDnsTxt");
exports.openAttestationDnsTxt = openAttestationDnsTxt_1.openAttestationDnsTxt;
var openAttestationEthereumDocumentStoreIssued_1 = require("./verifiers/documentStoreIssued/openAttestationEthereumDocumentStoreIssued");
exports.openAttestationEthereumDocumentStoreIssued = openAttestationEthereumDocumentStoreIssued_1.openAttestationEthereumDocumentStoreIssued;
var openAttestationEthereumDocumentStoreRevoked_1 = require("./verifiers/documentStoreRevoked/openAttestationEthereumDocumentStoreRevoked");
exports.openAttestationEthereumDocumentStoreRevoked = openAttestationEthereumDocumentStoreRevoked_1.openAttestationEthereumDocumentStoreRevoked;
var validator_1 = require("./validator");
exports.isValid = validator_1.isValid;
var openAttestationEthereumTokenRegistryMinted_1 = require("./verifiers/tokenRegistryMinted/openAttestationEthereumTokenRegistryMinted");
exports.openAttestationEthereumTokenRegistryMinted = openAttestationEthereumTokenRegistryMinted_1.openAttestationEthereumTokenRegistryMinted;
var openAttestationVerifiers = [
    openAttestationHash_1.openAttestationHash,
    openAttestationEthereumDocumentStoreIssued_1.openAttestationEthereumDocumentStoreIssued,
    openAttestationEthereumTokenRegistryMinted_1.openAttestationEthereumTokenRegistryMinted,
    openAttestationEthereumDocumentStoreRevoked_1.openAttestationEthereumDocumentStoreRevoked,
    openAttestationDnsTxt_1.openAttestationDnsTxt
];
exports.openAttestationVerifiers = openAttestationVerifiers;
var verify = verificationBuilder_1.verificationBuilder(openAttestationVerifiers);
exports.verify = verify;
__export(require("./types/core"));
__export(require("./types/error"));
