// @govtechsg/oa-verify/dist/esm/index.js
import { verificationBuilder } from "./verifiers/verificationBuilder";
import { openAttestationHash } from "./verifiers/hash/openAttestationHash";
import { openAttestationDnsTxt } from "./verifiers/dnsText/openAttestationDnsTxt";
import { openAttestationEthereumDocumentStoreIssued } from "./verifiers/documentStoreIssued/openAttestationEthereumDocumentStoreIssued";
import { openAttestationEthereumDocumentStoreRevoked } from "./verifiers/documentStoreRevoked/openAttestationEthereumDocumentStoreRevoked";
import { isValid } from "./validator";
import { openAttestationEthereumTokenRegistryMinted } from "./verifiers/tokenRegistryMinted/openAttestationEthereumTokenRegistryMinted";
var openAttestationVerifiers = [
    openAttestationHash,
    openAttestationEthereumDocumentStoreIssued,
    openAttestationEthereumTokenRegistryMinted,
    openAttestationEthereumDocumentStoreRevoked,
    openAttestationDnsTxt
];
var verify = verificationBuilder(openAttestationVerifiers);
export * from "./types/core";
export * from "./types/error";
export { verificationBuilder, openAttestationVerifiers, isValid, verify, openAttestationHash, openAttestationEthereumDocumentStoreRevoked, openAttestationEthereumDocumentStoreIssued, openAttestationDnsTxt, openAttestationEthereumTokenRegistryMinted };
