// @govtechsg/token-registry/src/contracts/ITitleEscrowCreatorFactory.ts
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */

import {Contract, Signer} from "ethers";
import {Provider} from "ethers/providers";

import {ITitleEscrowCreator} from "./ITitleEscrowCreator";

export class ITitleEscrowCreatorFactory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ITitleEscrowCreator {
    return new Contract(address, _abi, signerOrProvider) as ITitleEscrowCreator;
  }
}

const _abi = [
  {
    constant: false,
    inputs: [
      {
        internalType: "address",
        name: "tokenRegistry",
        type: "address"
      },
      {
        internalType: "address",
        name: "beneficiary",
        type: "address"
      },
      {
        internalType: "address",
        name: "holder",
        type: "address"
      }
    ],
    name: "deployNewTitleEscrow",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address"
      }
    ],
    payable: false,
    stateMutability: "nonpayable",
    type: "function"
  }
];
