// @govtechsg/token-registry/dist/contracts/HasHolderFactory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HasHolderFactory = void 0;

var _ethers = require("ethers");

/* Generated by ts-generator ver. 0.0.8 */

/* tslint:disable */
class HasHolderFactory {
  static connect(address, signerOrProvider) {
    return new _ethers.Contract(address, _abi, signerOrProvider);
  }

}

exports.HasHolderFactory = HasHolderFactory;
const _abi = [{
  inputs: [{
    internalType: "address",
    name: "_holder",
    type: "address"
  }],
  payable: false,
  stateMutability: "nonpayable",
  type: "constructor"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "previousHolder",
    type: "address"
  }, {
    indexed: true,
    internalType: "address",
    name: "newHolder",
    type: "address"
  }],
  name: "HolderChanged",
  type: "event"
}, {
  constant: true,
  inputs: [],
  name: "holder",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  payable: false,
  stateMutability: "view",
  type: "function"
}];
