// @govtechsg/token-registry/dist/contracts/MinterRoleFactory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MinterRoleFactory = void 0;

var _ethers = require("ethers");

/* Generated by ts-generator ver. 0.0.8 */

/* tslint:disable */
class MinterRoleFactory {
  static connect(address, signerOrProvider) {
    return new _ethers.Contract(address, _abi, signerOrProvider);
  }

}

exports.MinterRoleFactory = MinterRoleFactory;
const _abi = [{
  inputs: [],
  payable: false,
  stateMutability: "nonpayable",
  type: "constructor"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "account",
    type: "address"
  }],
  name: "MinterAdded",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "account",
    type: "address"
  }],
  name: "MinterRemoved",
  type: "event"
}, {
  constant: true,
  inputs: [{
    internalType: "address",
    name: "account",
    type: "address"
  }],
  name: "isMinter",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  payable: false,
  stateMutability: "view",
  type: "function"
}, {
  constant: false,
  inputs: [{
    internalType: "address",
    name: "account",
    type: "address"
  }],
  name: "addMinter",
  outputs: [],
  payable: false,
  stateMutability: "nonpayable",
  type: "function"
}, {
  constant: false,
  inputs: [],
  name: "renounceMinter",
  outputs: [],
  payable: false,
  stateMutability: "nonpayable",
  type: "function"
}];
